source/device_defender.c

```diff
@@ -269,6 +301,90 @@ static int s_get_metric_report_json(
+            goto cleanup;
+        }
+        cJSON_AddItemToObject(root, "custom_metrics", custom_metrics);
+
+        size_t list_size = 0;
+        struct cJSON *array_item = NULL;
+        struct cJSON *item = NULL;
+        struct cJSON *json_list = NULL;
+        struct cJSON *spurious_array_container = NULL;
+        for (size_t metric_index = 0; metric_index < custom_metrics_len; ++metric_index) {
+            spurious_array_container = cJSON_CreateArray();
+            if (NULL == spurious_array_container) {
+                goto cleanup;
+            }
+            cJSON_AddItemToObject(
+                custom_metrics,
+                aws_string_c_str(custom_metrics_data[metric_index].metric->metric_name),
+                spurious_array_container);
+
+            item = cJSON_CreateObject();
+            if (NULL == item) {
+                goto cleanup;
+            }
+            cJSON_AddItemToArray(spurious_array_container, item);
+
+            switch (custom_metrics_data[metric_index].metric->type) {
+                case DD_METRIC_NUMBER:
+                    cJSON_AddNumberToObject(item, "number", (double)custom_metrics_data[metric_index].data.number);
+                    break;
+                case DD_METRIC_NUMBER_LIST:
+                    list_size = aws_array_list_length(&custom_metrics_data[metric_index].data.list);
+                    json_list = cJSON_CreateArray();
+                    if (NULL == json_list) {
+                        goto cleanup;
+                    }
+                    cJSON_AddItemToObject(item, "number_list", json_list);
+                    for (size_t num_index = 0; num_index < list_size; ++num_index) {
+                        int64_t number = 0;
+                        aws_array_list_get_at(&custom_metrics_data[metric_index].data.list, &number, num_index);
+                        array_item = cJSON_CreateNumber((double)number);
+                        cJSON_AddItemToArray(json_list, array_item);
+                    }
+                    break;
+                case DD_METRIC_STRING_LIST:
+                    list_size = aws_array_list_length(&custom_metrics_data[metric_index].data.list);
+                    json_list = cJSON_CreateArray();
+                    if (NULL == json_list) {
+                        goto cleanup;
+                    }
+                    cJSON_AddItemToObject(item, "string_list", json_list);
+                    for (size_t string_index = 0; string_index < list_size; ++string_index) {
+                        struct aws_string *string_value = NULL;
+                        aws_array_list_get_at(
+                            &custom_metrics_data[metric_index].data.list, &string_value, string_index);
+                        array_item = cJSON_CreateString(aws_string_c_str(string_value));
+                        cJSON_AddItemToArray(json_list, array_item);
+                    }
+                    break;
+                case DD_METRIC_IP_LIST:
```

Comment:
It's modeled in DeviceDefender that way so that's out of our control. At the very least, we have to label it as such when generating the report even if we do nothing different otherwise. Internally, the callback function pointer type is different so it's somewhat sorted and typename visible.